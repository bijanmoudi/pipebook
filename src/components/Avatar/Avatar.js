import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { getAbbr } from '../../utils';
import './index.scss';

const Avatar = ({
	url,
	name,
	size,
	link,
	focusable,
	className,
	...restProps
}) => {
	const abbr = name ? getAbbr(name) : null;
	const WrapperElement = link ? 'a' : 'div';
	const [loadedURL, setURL] = useState(null);
	const loaderImg = new Image();
	const setImageURL = () => {
		setURL(url);
	};
	useEffect(() => {
		if (url) {
			loaderImg.src = url;
			loaderImg.addEventListener('load', setImageURL);
			return () => {
				loaderImg.removeEventListener('load', setImageURL);
			};
		}
	});
	return (
		<figure
			className={classnames(
				'avatar',
				{ 'avatar--is-loading': !loadedURL && url },
				{ [`avatar--${size}`]: size },
				{ [className]: className }
			)}
			{...restProps}
		>
			<WrapperElement
				title={name}
				className="avatar__wrapper"
				{...(link ? { href: link } : {})}
				{...(link && !focusable ? { tabIndex: '-1' } : {})}
			>
				{abbr ? <abbr className="avatar__abbreviation">{abbr}</abbr> : null}
				{url && loadedURL ? (
					<div
						role="img"
						aria-label={`${name}'s Avatar`}
						className="avatar__image"
						style={{
							backgroundImage: `url(${loadedURL})`
						}}
					/>
				) : null}
			</WrapperElement>
		</figure>
	);
};

Avatar.propTypes = {
	url: PropTypes.oneOfType([() => null, PropTypes.string]),
	name: PropTypes.oneOfType([() => null, PropTypes.string]),
	link: PropTypes.oneOfType([() => null, PropTypes.string]),
	size: PropTypes.oneOfType([
		() => null,
		PropTypes.oneOf(['large', 'default'])
	]),
	focusable: PropTypes.bool
};

Avatar.defaultProps = {
	url: null,
	name: null,
	link: null,
	size: null,
	focusable: false
};

export default Avatar;
