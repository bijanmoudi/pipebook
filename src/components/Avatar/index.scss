/*=====================================
=            Avatar Styles            =
=====================================*/

@import '../../assets/styles/constants/layout';
@import '../../assets/styles/helpers/colors';
@import '../../assets/styles/helpers/units';

.avatar {
	position: relative;
	display: inline-block;
	vertical-align: middle;
	width: em(50);
	height: em(50);
	user-select: none;
	&--large {
		font-size: em(29);
	}
	&__image,
	&__wrapper {
		display: block;
		width: 100%;
		height: 100%;
	}
	&__wrapper {
		border-radius: 100%;
		overflow: hidden;
		background-color: get-color('blue-light');
		color: get-color('blue');
		font-weight: $semi-bold-font-weight;
		text-align: center;
		line-height: em(50);
		text-decoration: none;
		&,
		&:active {
			box-shadow: 0 0 0 rem(1) get-color('gray-light-very');
		}
		&:active {
			transform: scale(1);
		}
	}
	&__image {
		position: absolute;
		top: 0;
		left: 0;
		border-radius: 100%;
		opacity: 1;
		background-repeat: no-repeat;
		background-position: center center;
		background-size: cover;
		transition: opacity 250ms ease 500ms;
	}
	&--is-loading {
		.avatar__image {
			opacity: 0;
		}
	}
}
