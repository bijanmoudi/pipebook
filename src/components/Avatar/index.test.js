import React from 'react';
import renderer from 'react-test-renderer';
import Avatar from './Avatar';

const props = {
	url: 'http://placehold.it/200x200',
	name: 'Dulce Mays',
	link: 'https://www.google.com/search?q=Dulce+Mays'
};

describe('Testing "Avatar" component', () => {
	it('Rendering without image', () => {
		const tree = renderer.create(<Avatar name={props.name} />).toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('Rendering without name', () => {
		const tree = renderer.create(<Avatar url={props.url} />).toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('Rendering with image, name, and link', () => {
		const tree = renderer
			.create(
				<Avatar name={props.name} url={props.url} link={props.link} />
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
