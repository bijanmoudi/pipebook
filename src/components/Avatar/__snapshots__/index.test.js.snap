// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing "Avatar" component Rendering with image, name, and link 1`] = `
<figure
  className="avatar avatar--is-loading"
>
  <a
    className="avatar__wrapper"
    href="https://www.google.com/search?q=Dulce+Mays"
    tabIndex="-1"
    title="Dulce Mays"
  >
    <abbr
      className="avatar__abbreviation"
    >
      DM
    </abbr>
    <div
      aria-label="Dulce Mays's Avatar"
      className="avatar__image"
      role="img"
      style={
        Object {
          "backgroundImage": "url(null)",
        }
      }
    />
  </a>
</figure>
`;

exports[`Testing "Avatar" component Rendering without image 1`] = `
<figure
  className="avatar"
>
  <div
    className="avatar__wrapper"
    title="Dulce Mays"
  >
    <abbr
      className="avatar__abbreviation"
    >
      DM
    </abbr>
  </div>
</figure>
`;

exports[`Testing "Avatar" component Rendering without name 1`] = `
<figure
  className="avatar avatar--is-loading"
>
  <div
    className="avatar__wrapper"
    title={null}
  >
    <div
      aria-label="null's Avatar"
      className="avatar__image"
      role="img"
      style={
        Object {
          "backgroundImage": "url(null)",
        }
      }
    />
  </div>
</figure>
`;
